### ISSUE TITLE
Interactor composition

### ISSUE THREAD

**Issue by cowboyd** [thefrontside/bigtest#426]
Interactors (mostly) are used to specify and activate components. The idea is that in your test case, you will load up the app with the `App` interactor, and the click buttons, move pointers, type fields etc. in order to push the state of the application forward just as would a user:

```ts
import { Button, Input } from '@bigtest/interactor';

test("login")
  .step("load the app", App.visit('/signin'))
  .step("fill in username", Input("username").fillIn('cowboyd'))
  .step("fill in password", Input("password").fillIn('open sesame'))
  .step("submit the form", Button("submit").click())
```

We've pretty much come around to the view that we want to treat interactors as units that are defined by _what they can do_ rather than _what they contain_. This comes into play when we want to write interactors for more complex components than just simple html elements. Let's use a date picker interactor for example. The datepicker could have an input field to enter text directly, you could use a popup to select the text. Let's imagine we'd use it like:

```ts
step("enter CC expiration", DatePicker("Expiration Date").enter("062020"))
```

What we're specfically rejecting is the idea that the expiration date field is exposed via the datepicker, so you would not see the following construction as you would in original interactors.

```ts
step("enter CC expiration", DatePicker("Expiration Date").input.fillIn("062020"))
```

Instead, interactors are like components: once you draw the line around them, the only way you can interact with them is to call specific actions. However, the fact remains that the date picker itself, _is, in fact, composed of a field, button and popup_ so how do we re-use the interactors for those elements? One apparent choice is that composition must happen within actions.

This would mean that somehow we want to look up the field within `DatePicker` and then call the `fillIn` method on that. We might imagine writing it like this:

```ts
export const DatePicker = createInteractor<HTMLDivElement>("DatePicker")({
  selector: 'div',
  defaultLocator: element => element.title,
  actions: {
    enter(element, value: string) { element.querySelector('input:first').value = value }
  }
})
```

That might work, but what we really we want to use the `Input().fillIn()` interaction that we used up above. What if, for example, we changed the logic of the `fillIn()` from setting the `.value` directly to actually focusing, dispatching individual key events, and then bluring? We'd want to pick up that improvement without having to change our datepicker interactor at all. What we'd really like to say is something like:

```ts
export const DatePicker = createInteractor<HTMLDivElement>("DatePicker")({
  selector: 'div',
  defaultLocator: element => element.title,
  actions: {
    enter: (picker, value) => picker.find('input:first').fillIn(value)
  }
})
```

`picker` in this case is the actual datepicker interactor which can then do a switcheroo on the action that actually gets run. It's almost like we need some sort of monadic actions that either do the thing, or return another set of things that does the thing.

composing actions themselves
-------------------------------

The discussion so far has been about how to compose interactors that are part of the private API of another interactor at the action level. In other words, we don't want to expose the internal structure of the datepicker to consumers of the datepicker interactor. Instead, we want to treat it as if it were its own html element. 

However, the other question of composition that we haven't really tackled is how to model *sequences of interactions* because ultimately, we'll need to not just abstract away DOM composition, but also create a single action out of many. To use our datepicker example again, what if we wanted to model the selection of a date via the popup UI? This might involve

1. click the choose button
2. selecting the year from the popup drop down
3. selecting the month from the popup drop down
4. selecting the day from the popup
5. hit the select button

Individually, these interactions are:

```ts
Button("choose").click();
Popup("calendar").find(Select("year").select("2020"));
Popup("calendar").find(Select("month").select("Jun"));
Popup("calendar").find(Button("31")).click();
```

So how do we stitch these together into a single interaction?

**Comment by patrickberkeley** [comment]
Regarding composition – and without knowing much about this – could an interactor consume interactors? E.g.,

```ts
import { Button, Select } from '@bigtest/interactor';

export const DatePicker = createInteractor<HTMLDivElement>("DatePicker")({
  selector: 'div',
  defaultLocator: element => element.title,
  actions: {
    enter(picker, value: string) {
      const { month, date, year } = parseDate(value);

      picker.find(Select("year").select(year));
      picker.find(Select("month").select(month));
      picker.find(Button(date)).click();
    }
  }
})
```

**Comment by patrickberkeley** [comment]
One other question I had about the above `App.visit()` as it relates to

> we want to treat interactors as units that are defined by _what they can do_ rather than _what they contain_

So could there be a `Visit` interactor instead of `App.visit()`?

**Comment by cowboyd** [comment]
> One other question I had about the above `App.visit()` as it relates to
> 
> > we want to treat interactors as units that are defined by _what they can do_ rather than _what they contain_
> 
> So could there be a `Visit` interactor instead of `App.visit()`?

Absolutely, this would be a pretty straightforward change. What would be the difference in your mind?

**Comment by cowboyd** [comment]
> Regarding composition – and without knowing much about this – could an interactor consume interactors? E.g.,

I think that this is exactly what we want to do: inside actions, use the current context plus other interactors to specify interactions. The problem is how to do so in a way that holds up the guarantees that interactors provide. In your example, these statements;

```ts
picker.find(Select("year").select(year));
picker.find(Select("month").select(month));
picker.find(Button(date)).click();
```

Have no relation to each other and will be evaluated in the same tick of the run loop, so how do we specify that _first_ select the year, _then_ select month, and _then_ click date?

we could use an async function somehow, but we also want to surface the fact that these are the three things that happened, so that if an error occurs, we know exactly where it happened. All this is to say that we need to thread that context through each of these separate actions so that the runtime can recognize them as not separate at all, but as part of a larger continuum of execution.

It's a real head-scratcher.

**Comment by patrickberkeley** [comment]
> > One other question I had about the above `App.visit()` as it relates to
> > > we want to treat interactors as units that are defined by _what they can do_ rather than _what they contain_
> > 
> > 
> > So could there be a `Visit` interactor instead of `App.visit()`?
> 
> Absolutely, this would be a pretty straightforward change. What would be the difference in your mind?

`App.visit()` seems to be more _what they contain_ whereas `Visit()` seems more _what they can do_. So the latter seems more aligned with the goal, and analogous to the other interactors.

**Comment by cowboyd** [comment]
@patrickberkeley that's a great point. The thing is, there is only ever a single app, so really it's just a namespace for "singleton" interactors and the actions that act on the implicit state of a webpage. What if we made it just a free-form function:

```js
import { visit } from '@bigtest/interactor';
```

**Comment by taras** [comment]
@cowboyd technically, a single app that is a component could become one component of many others. Maybe visit should be referring to the Browser rather than the app?

**Comment by patrickberkeley** [comment]
Regarding composition:

> use an async function

That was my first thought based on your reply. 

>  if an error occurs, we know exactly where it happened.

Would `try/catch` blocks or `.then().catch()` work? The thrown error could aggregate some key from each method in the chain. I'm having a hard time figuring out async chaining, but something like this which would ideally be a fully typed class and fancier error formatting:

```ts
const trace: Array<string> = [];

const Chainable = (promise = Promise.resolve ()) => (
  {
    a: () => {
      trace.push('A');
      const fn = promise
        .then(() => {
          console.log('a');
        })
        .catch(() => {
          throw(trace);
        });
      return Chainable(fn);
    },
    b: () => {
      trace.push('B');
      const fn = promise
        .then(() => {
          console.log('b');
        })
        .catch(() => {
          throw(trace);
        });
      return Chainable(fn);
    },
    c: () => {
      trace.push('C');
      const fn = promise
        .then(() => {
          Promise.reject(trace);
          console.log('c');
        })
        .catch(() => {
          throw(trace);
        });
      return Chainable(fn);
    },
  }
);

Chainable().a().b().c();
```

**Comment by cowboyd** [comment]
Just noting for posterity on why this use-case is evidence for why `async/await` was a missed opportunity. Because it is a fundamentally closed and finite context (that of promise resolution), whereas generators have an open and infinite context.

I'm musing whether they could be used here:

```ts
yield selectYear();
yield selectMonth();
yield saveDate();
```

The nice thing here is that we can capture this sequence of operations and do whatever we want with them.

**Comment by jnicklas** [comment]
This was fixed by #451
